Add documentation about JLL packages (#772)
JuliaPackaging/BinaryBuilder.jl: docs/src/jll.md

# JLL packages

`BinaryBuilder.jl` is designed to produce tarballs that can be used in any
environment, but so far their main use has been to provide pre-built libraries
and executables to be readily used in Julia packages.  This is accomplished by
JLL packages (a pun on "Dynamic-Link Library", with the J standing for Julia).
They can be installed like any other Julia packages with the [Julia package
manager](https://julialang.github.io/Pkg.jl/v1/) in the REPL with
```
]add NAME_jll
```
and then loaded with
```
using NAME_jll
```
However, most users will not ever need to do these steps on their own, JLL

> packages are usually only used as dependency of packages wrapping binary

Reviewer: ```suggestion
packages are usually only used as dependencies of packages wrapping binary
```

  shared libraries for the current platform.  This is equal to `LD_LIBRARY_PATH`
  on Linux and FreeBSD, `DYLD_FALLBACK_LIBRARY_PATH` on macOS, and `PATH` on
  Windows;

The wrapper files for each platform also define the
[`__init__()`](https://docs.julialang.org/en/v1/manual/modules/index.html#Module-initialization-and-precompilation-1)
function of the JLL package, the code that is executed every time the package is
loaded.  The `__init__()` function will populate most of the variables mentioned
above and automatically open the shared libraries, if any, listed in the
products of the `build_tarballs.jl` script that generated the JLL package.

The rest of the code in the wrappers is specific to each of the products of the
JLL package and detailed below.  If you want to see a concrete example of a
package providing all the main three products, have a look at
[`Fontconfig_jll.jl`](https://github.com/JuliaBinaryWrappers/Fontconfig_jll.jl/tree/785936d816d1ae65c2a6648f3a6acbfd72535e36).

### LibraryProduct

A [`LibraryProduct`](@ref) is a shared library that can be
[`ccall`](https://docs.julialang.org/en/v1/manual/calling-c-and-fortran-code/)ed

> from Julia.  Assuming that the product is called `libdataproc`, te wrapper

Reviewer: ```suggestion
from Julia.  Assuming that the product is called `libdataproc`, the wrapper
```

defines the following variables:

* `libdataproc`: this is the exported
  [`const`](https://docs.julialang.org/en/v1/manual/variables-and-scoping/#Constants-1)
  variable that should be used in a `ccall`.  To make it simply, the value of
  this variable is roughly the basename of the shared library;
* `libdataproc_splitpath`: the path of the shared library, relative to
  `artifact_dir`, as returned by
  [`splitpath`](https://docs.julialang.org/en/v1/base/file/#Base.Filesystem.splitpath);
* `libdataproc_path`: the full absolute path of the shared library;
* `libdataproc_handle`: the address in memory of the shared library after it has
  been loaded at initialization time.

### ExecutableProduct

An [`ExecutableProduct`](@ref) is a binary executable that can be run on the
current platform.  If, for example, the `ExecutableProduct` has been called
`mungify_exe`, the wrapper defines an exported function named `mungify_exe` that
should run by the user in the following way:

```julia
mungify_exe() do exe

>     run(`exe $arguments`)

Reviewer: ```suggestion
    run(`$exe $arguments`)
```

end
```

Note that in this example `exe` can be replaced with any name of your choice:
with the
[`do`-block](https://docs.julialang.org/en/v1/manual/functions/#Do-Block-Syntax-for-Function-Arguments-1)
syntax you are defining the name of the variable that will be used to actually
call the binary with
[`run`](https://docs.julialang.org/en/v1/base/base/#Base.run).

A common point of confusion about `ExecutableProduct`s in JLL packages is why
this function is needed: while in principle you could directly run the
executable directly by using its absolute path in `run`, this wrapper function

> ensures that the executable will find the shared library it needs while running.

Reviewer: ```suggestion
ensures that the executable will find all shared libraries it needs while running.
```
